keps/sig-scheduling/20190317-non-preempting-priorityclass

```diff
@@ -0,0 +1,104 @@
+---
+title: Add NonPreempting Option For PriorityClasses
+authors:
+  - "@vllry"
+owning-sig: sig-scheduling
+participating-sigs:
+  - sig-scheduling
+reviewers:
+  - "k82cn"
+  - "wgliang"
+approvers:
+  - "bsalamat"
+editor: Vallery Lancey
+creation-date: 2019-03-17
+last-updated: 2019-03-17
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Allow PriorityClasses To Be Non-Preempting
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+
+## Summary
+
+[PriorityClasses](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/) are a beta feature,
+which impact the scheduling and eviction of pods.
+Pods will be scheduled according to descending priority.
+If a pod cannot be scheduled due to insufficient resources,
+lower-priority pods will be descheduled to make room.
+
+This proposal makes the pre-empting (descheduling) behavior optional,
+by adding a new field to PriorityClasses.
+If a PriorityClass does not have preemption enabled,
+the scheduler will not preempt pods in order to schedule a pod of that priority.
+
+## Motivation
+
+High-priority, non-preempting workloads are a common data science use case.
+Preempting batch workloads is a waste, as the work unit must be repeated.
+
+### Goals
+
+Add a boolean to PriorityClasses,
+to enable or disable preemption for pods of that PriorityClass.
+
+### Non-Goals
+
+## Proposal
+
+Add a NonPreempting field to PriorityClasses.
+This field will default to false,
+for backwards compatibility.
+
+If NonPreempting is false,
+the scheduler will preempt lower priority pods to schedule this pod,
+as is current behavior.
+
+If NonPreempting is true,
```

Comment:
Do we support update this?

Reply:
The nice thing about this KEP is that it only affects the scheduling of new pods and not the "evictability" of running pods. Updating this value seems simple to support.

Reply:
hmm.... if we add a new field in `PodTemplate`, it's hard to update created pods :) If we hold PriorityClass in scheduler, how about kubelet?

Reply:
I'll look more into this.

Reply:
Looks like we're going with denormalizing the field into PodSpec... so from my understanding, that will present some upgrade challenges.